### PR TITLE
Replicate can specify blocks to copy

### LINKED ISSUE #2713: Feature Request: Replicate can copy individual blocks
So today we faced an issue where a few blocks were not copied to our destination bucket and I had to manually copy them over, I am still trying to figure out what went wrong but it would be a nice feature to have if the replicate can copy individual blocks over.

One another feature that would be really useful is to verify if all the blocks have been copied over between 2 buckets and maybe give a report for those which might have been missed.

### PR DESCRIPTION
**Pull request by yeya24**
Signed-off-by: Ben Ye <yb532204897@gmail.com>

* [x] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Changes

Fixes #2713 

Added a repeated flag `--id` to specify the blocks to replicate.

### CODE REVIEW

File: pkg/replicate/replicator.go [hunk 1]
```diff
@@ -83,6 +83,7 @@ func RunReplicate(
 	toObjStoreConfig *extflag.PathOrContent,
 	singleRun bool,
 	minTime, maxTime *thanosmodel.TimeOrDurationValue,
+	blockIDs []string,
```
Comment (bwplotka) [inline pkg/replicate/replicator.go:86]: why not pasing `ulid.ULID`?

File: pkg/replicate/scheme.go [hunk 2]
```diff
@@ -64,6 +71,16 @@ func (bf *BlockFilter) Filter(b *metadata.Meta) bool {
 		return false
 	}
 
+	// If required block IDs are set, we only match required blocks and ignore others.
+	if len(bf.blockIDs) > 0 {
+		id := b.ULID.String()
+		if _, ok := bf.blockIDs[id]; ok {
+			return true
+		}
+		level.Debug(bf.logger).Log("msg", "filtering block", "reason", "block ID doesn't match", "block_uuid", id)
```
Comment (bwplotka) [inline pkg/replicate/scheme.go:80]: This will spam a lot. Maybe we can just not do log it?

File: pkg/replicate/scheme.go [hunk 3]
```diff
@@ -49,11 +51,16 @@ func NewBlockFilter(
 	for _, compactionLevel := range compactionLevels {
 		allowedCompactions[compactionLevel] = struct{}{}
 	}
+	blockSet := make(map[string]struct{})
```
Comment (bwplotka) [inline pkg/replicate/scheme.go:54]: The blockID list will be super short. Can we just loop over it instead of using map? Might be simpler (:

**Review by bwplotka — COMMENTED** [review]
Nice, job! Some suggestions :hugs:

File: pkg/replicate/scheme.go (hunk 3 above)
Comment (yeya24) [reply pkg/replicate/scheme.go:54]: Sounds better!

**Comment by yeya24** [comment]
@bwplotka Updated, please take a look again!

File: cmd/thanos/tools_bucket.go [hunk 4]
```diff
@@ -459,6 +460,15 @@ func registerBucketReplicate(app extkingpin.AppClause, objStoreConfig *extflag.P
 			resolutionLevels = append(resolutionLevels, compact.ResolutionLevel(lvl.Milliseconds()))
 		}
 
+		blockIDs := make([]ulid.ULID, 0, len(*ids))
+		for _, id := range *ids {
+			if bid, err := ulid.Parse(id); err != nil {
+				return errors.Wrap(err, "invalid ULID found in --id flag")
+			} else {
```
Comment (bwplotka) [inline cmd/thanos/tools_bucket.go:467]: no need for else ;p

File: pkg/compactv2/modifiers_test.go [hunk 5]
```diff
@@ -0,0 +1,33 @@
+package compactv2
+
+import (
+	"github.com/prometheus/prometheus/tsdb/tombstones"
+	"reflect"
+	"testing"
+)
+
+func TestIntersection(t *testing.T) {
```
Comment (bwplotka) [inline pkg/compactv2/modifiers_test.go:9]: Unrelated?

**Review by bwplotka — APPROVED** [review]
LGTM, just two nits (:

File: pkg/compactv2/modifiers_test.go (hunk 5 above)
Comment (yeya24) [reply pkg/compactv2/modifiers_test.go:9]: Sorry, will remove this.

**Comment by yeya24** [comment]
Updated again.

**Review by bwplotka — APPROVED** [review]
LGTM :muscle: rebase only.

**Comment by yeya24** [comment]
> LGTM  rebase only.

Rebased, PTAL

**Review by GiedriusS — APPROVED** [review]
LGTM, thank you! Please rebase on `master` again and I think we can merge this.

**Comment by yeya24** [comment]
Done. PTAL again.

**Comment by yeya24** [comment]
Just rebased this pr. PTAL again, thanks!

**Comment by yeya24** [comment]
I will merge this since there are already two approvals!
